torchtune/models/t5/_encoder.py

```diff
@@ -0,0 +1,304 @@
+
+
+class T5EncoderSelfAttention(nn.Module):
+    """
+    Self-attention for the T5 encoder.
+
+    Standard self-attention with two differences:
+        - No scaling factor
+        - Add "relative position bias" to the attention scores.
+            (See: :class:`~torchtune.models.t5._encoder.T5EncoderRelativePositionBias`)
+
+    Args:
+        embed_dim (int): The model dimension.
+        num_heads (int): Number of attention heads.
+        head_dim (int): Dimension of the attention heads (should equal `embed_dim // num_heads`)
+        q_proj (nn.Module): Projection layer for query.
+        k_proj (nn.Module): Projection layer for key.
+        v_proj (nn.Module): Projection layer for value.
+        output_proj (nn.Module): Projection layer for output.
+
+    Raises:
+        ValueError: If ``num_heads % num_kv_heads != 0``
+        ValueError: If ``embed_dim // num_heads != head_dim``
+    """
+
+    def __init__(
+        self,
+        embed_dim: int,
+        num_heads: int,
+        head_dim: int,
+        q_proj: nn.Module,
+        k_proj: nn.Module,
+        v_proj: nn.Module,
+        output_proj: nn.Module,
+    ):
+        super().__init__()
+        if embed_dim % num_heads != 0:
+            raise ValueError(
+                f"embed_dim ({embed_dim}) must be divisible by "
+                f"num_heads ({num_heads})"
+            )
+        if embed_dim // num_heads != head_dim:
+            raise ValueError(
+                f"head_dim ({head_dim}) must be equal to embed_dim // num_heads"
+            )
+
+        self.num_heads = num_heads
+        self.head_dim = head_dim
+
+        self.q_proj = q_proj
+        self.k_proj = k_proj
+        self.v_proj = v_proj
+        self.output_proj = output_proj
+
+    def forward(self, x: Tensor, rel_pos_bias: Tensor) -> Tensor:
+        """
+        Args:
+            x (Tensor): input tensor with shape [bsz, seq_len, embed_dim]
+            rel_pos_bias (Tensor): relative position bias with shape [1, num_heads, max_seq_len, max_seq_len]
+                See: :class:`~torchtune.models.t5._encoder.T5EncoderRelativePositionBias`
+
+        Returns:
+            Tensor: output tensor with shape [bsz, seq_len, embed_dim]
+        """
+        bsz, seq_len, embed_dim = x.shape
+
+        # QKV projections
+        q = self.q_proj(x)
+        k = self.k_proj(x)
+        v = self.v_proj(x)
+
+        # [bsz, seq_len, embed_dim] -> [bsz, num_heads, seq_len, head_dim]
+        q = q.view(bsz, seq_len, self.num_heads, self.head_dim).transpose(1, 2)
+        k = k.view(bsz, seq_len, self.num_heads, self.head_dim).transpose(1, 2)
+        v = v.view(bsz, seq_len, self.num_heads, self.head_dim).transpose(1, 2)
+
+        # attention with relative position bias
+        attn_score = torch.matmul(q, k.transpose(-2, -1))
+        attn_score += rel_pos_bias
+        attn_weight = F.softmax(attn_score.float(), dim=-1).to(attn_score.dtype)
+        attn_out = torch.matmul(attn_weight, v)
```

Comment:
This part could be simplified by using `F.scaled_dot_product_attention` by repurposing the `mask` argument for `rel_pos_bias` (because `scaled_dot_product_attention` simply adds the mask to the attention score when the mask is a float tensor). However, when I tried this it was significantly slower for some reason

Reply:
I wonder whether this is a case where we could benefit from using flex attention?

Reply:
Flex attention requires boolean masks AFAICT: https://github.com/pytorch/pytorch/blob/main/torch/nn/attention/flex_attention.py#L826-L827

Reply:
I believe you can use any mask for flexattention with scoremod. I think we should land this as is but create an issue to update this to use flex attention.